src/main.js

```diff
@@ -0,0 +1,501 @@
+          ready: new DeferredPromise(),
+        };
+        this.promises.GeoTIFF.then((tiff) => {
+          self.GeoTIFF = tiff;
+          return tiff.getImageCount();
+        })
+          .then((count) => {
+            self.imageCount = count;
+            let promises = [...Array(count).keys()].map((index) => self.GeoTIFF.getImage(index));
+            return Promise.all(promises);
+          })
+          .then((images) => {
+            self.GeoTIFFImages = images;
+            self.promises.GeoTIFFImages.resolve(images);
+            this.setupLevels();
+          })
+          .catch((error) => {
+            console.error("Re-throwing error with GeoTIFF:", error);
+            throw error;
+          });
+      }
+    }
+
+    static getAllTileSources = async (input, opts) => {
+      const fileExtension =
+        input instanceof File ? input.name.split(".").pop() : input.split(".").pop();
+
+      let tiff = input instanceof File ? fromBlob(input) : fromUrl(input);
+
+      return tiff
+        .then((t) => {
+          tiff = t;
+          return t.getImageCount();
+        })
+        .then((c) =>
+          Promise.all([...Array(c).keys()].map(async (index) => (await tiff).getImage(index)))
+        )
+        .then((images) => {
+          // Filter out images with photometricInterpretation.TransparencyMask
+          images = images.filter(
+            (image) =>
+              image.fileDirectory.photometricInterpretation !==
+              globals.photometricInterpretations.TransparencyMask
+          );
+
+          // Sort by width (largest first), then detect pyramids
+          images.sort((a, b) => b.getWidth() - a.getWidth());
+
+          // find unique aspect ratios (with tolerance to account for rounding)
+          const tolerance = 0.015;
+
+          const aspectRatioSets = images.reduce((accumulator, image) => {
+            const r = image.getWidth() / image.getHeight();
+            const exists = accumulator.filter(
+              (set) => Math.abs(1 - set.aspectRatio / r) < tolerance
+            );
+            if (exists.length === 0) {
+              let set = {
+                aspectRatio: r,
+                images: [image],
+              };
+              accumulator.push(set);
+            } else {
+              exists[0].images.push(image);
+            }
+            return accumulator;
+          }, []);
+
+          const imageSets = aspectRatioSets.map((set) => set.images);
+
+          return imageSets.map((images, index) => {
+            // Check if QPTIFF
+            if (index !== 0) {
+              return new OpenSeadragon.GeoTIFFTileSource(
+                {
+                  GeoTIFF: tiff,
+                  GeoTIFFImages: images,
+                },
+                opts
+              );
+            }
+
+            switch (fileExtension) {
+              case "qptiff":
+                const channels = parsePerkinElmerChannels(images);
```

Comment:
Ah, I see this is where this is used.

Reply:
Yup this is where I feel it's not clean yet - since I suddenly introduce a hardcoded identifier. I was thinking of adding a specific TIFFReader class that handles the image setup